### ISSUE TITLE
gdb.error 'Cannot find thread-local storage'

### ISSUE THREAD

**Issue by conceptofproof** [cloudburst/libheap#24]
I installed via the instructions in the installation guide and also changed `~/.local/lib/python3.5/site-packages/libheap/libheap.cfg` to `2.23`, which is my installed libc version, but I still get this issue:

```
gdb-peda$ heapls
[!] No gdb frame is currently selected.

Python Exception <class 'gdb.error'> Cannot find thread-local storage for process 20952, shared library /lib/x86_64-linux-gnu/libc.so.6:
Cannot find thread-local variables on this target: 
Error occurred in Python command: Cannot find thread-local storage for process 20952, shared library /lib/x86_64-linux-gnu/libc.so.6:
Cannot find thread-local variables on this target
```

```
➜  /home apt-cache policy libc6
libc6:
  Installed: 2.23-0ubuntu7
  Candidate: 2.23-0ubuntu7
  Version table:
 *** 2.23-0ubuntu7 500
        500 http://us.archive.ubuntu.com/ubuntu xenial-updates/main amd64 Packages
        500 http://security.ubuntu.com/ubuntu xenial-security/main amd64 Packages
        100 /var/lib/dpkg/status
     2.23-0ubuntu3 500
        500 http://us.archive.ubuntu.com/ubuntu xenial/main amd64 Packages
```
I am running this on Ubuntu 16.04 Xenial 64-bit.

Any suggestions on how I can fix this?

**Comment by cloudburst** [comment]
Can you try this command inside your gdb-peda:

```
gdb-peda$ python print(gdb.selected_frame().read_var("main_arena"))
```

That error message is propagating up from my API over the gdb python `gdb.selected_frame()`.  I'm not really sure why that error would be occurring.

Once you are attached to a process you should be able to see this API work via something like:

`gdb-peda$ python print(gdb.selected_frame())`

I'm still on 16.04.2 LTS and its what I use libheap on the most and that doesn't occur here to debug unfortunately.  I'm not sure what's up with that thread-local storage error either.  I'd guess GDB doesn't like the way you are spawning/attaching to a process for some reason?

```
bash$ apt-cache policy libc6
libc6:
  Installed: 2.23-0ubuntu7

bash$  apt-cache policy gdb
gdb:
  Installed: 7.11.1-0ubuntu1~16.04
```

**Comment by cloudburst** [comment]
Please reopen if there is some more info I could use to help reproduce this.

**Comment by zh-explorer** [comment]
hello, I am a user for libheap. And I meet the same problem in my ubuntu16.04.2 TLS.
I use libheap well in my old ubuntu14.04 TLS, so I use the same version libheap do some little test and I think I find how this happen.
I my old ubuntu with this gdb version :
```
apt-cache policy gdb 
gdb:
  Installed: 7.11.1-0ubuntu1~16.04
```
I try `gdb-peda$ python print(gdb.selected_frame().read_var("thread_arena"))`
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ValueError: Variable 'thread_arena' not found.
Error while executing Python code.
```
The gdb raise a ValueError to show that thread_arena is not exist.This exception whill be ignored

but in my new ubuntu16.04 with this gdb version
```
gdb:
  Installed: 7.11.1-0ubuntu1~16.04
```
the result of same cmd is
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
gdb.error: Cannot find thread-local storage for process 14948, shared library /lib/x86_64-linux-gnu/libc.so.6:
Cannot find thread-local variables on this target
Error while executing Python code.
```
And code catch this RuntimeError and end.

I'm still work to find out either this is a new change whit the updata of my gdb or there is something wrong in my gdb

**Comment by zh-explorer** [comment]
I try gdb 7.12 in my ubuntu and try defult gdb in ubuntu 17.04/16.04 use docker.
It seems that the new version of gdb change is behavior and make "read of thread_arena" as a RuntimeError

**Comment by cloudburst** [comment]
@zh-explorer thank you for the information!

I wonder if the GDB Python API changed.

**Comment by wbowling** [comment]
This was happening for me with single threaded processes, multithreaded applications work as expected. 

I think https://github.com/cloudburst/libheap/blob/master/libheap/pydbg/pygdbpython.py#L137 was  returning a different exception than `ValueError` when looking up `thread-local`.

**Comment by Kyle-Kyle** [comment]
Same problem here. Here is the error report.
~~~~
[!] No gdb frame is currently selected.

Python Exception <class 'gdb.error'> Cannot find thread-local storage for process 10109, shared library /lib/x86_64-linux-gnu/libc.so.6:
Cannot find thread-local variables on this target: 
Error occurred in Python command: Cannot find thread-local storage for process 10109, shared library /lib/x86_64-linux-gnu/libc.so.6:
Cannot find thread-local variables on this target
~~~~
And here is the debug output
~~~~
(gdb) python print(gdb.selected_frame())
{stack=0x7fffffffe490,code=0x4005f6,!special}
~~~~

**Comment by cq674350529** [comment]
First, thanks for your great job @cloudburst .

 I met the same problem, as @wbowling mentioned, line 137 in `pydbg/pygdbpython.py` was generating an different exception than `ValueError`. By debugging, I found a possible solution: try to catch `RuntimeError` before `ValueError` in line 137. 

It works well now, althougth I don't know what effects it may have.

**Comment by Kyle-Kyle** [comment]
Thank you for your advice.@cq674350529 
But there are still some problems:
First: The warnings are still there
~~~~
(gdb) fastbins
[!] No gdb frame is currently selected.
~~~~

Second: New problem. Occur only if there is a long list.
~~~~
---Type <return> to continue, or q <return> to quit---q
Python Exception <class 'KeyboardInterrupt'> Quit: 
Error occurred in Python command: Quit
~~~~
And after that, what I input will be colored

**Comment by cq674350529** [comment]
Year, the warning still exists. As @wbowling mentioned, even in a single threaded process, [line 48 in heapls.py](https://github.com/cloudburst/libheap/blob/master/libheap/frontend/commands/gdb/heapls.py#48) will read variable from `thread_arena`, so it outputs the warning. But for multithreaded applications , it's ok.  In my opinion, the warning doesn't make sense.

As for the second problem, I haven't met it yet.  You can give me(cq674350529@163.com) some additional imformation for reproducing it if you like.

**Comment by HolmesSherlock** [comment]
On Ubuntu 16.04.2, libc-2.23; I am facing the same error with the latest version of libheap checked out.

**Comment by cq674350529** [comment]
@HolmesSherlock Did you try to catch RuntimeError by modifying the code  as I suggested above? The latest version of libheap may not be updated.

**Comment by Kira-cxy** [comment]
Ubuntu 16.04.3, libc-2.23

**Comment by Kira-cxy** [comment]
Ubuntu 16.04.3, libc-2.23
Same problem here.

**Comment by mdu4rt3** [comment]
How can we make this work again for single threaded applications? Any ideas?

Ubuntu 16.04.3, libc-2.23

**Comment by cloudburst** [comment]
@mdu4rt3 You can either revert 59a3d38a63cc4a17a50d61fa20efe176f2887b28 or use one of the manual patches earlier in this issue.

**Comment by neversayevern** [comment]
I am getting the same Cannot find thread-local storage error. 

$ apt-cache policy libc6
```
libc6:
  Installed: 2.24-17
  Candidate: 2.24-17
  Version table:
```

$ cat /home/pyxmr/.local/lib/python2.7/site-packages/libheap/libheap.cfg 
```
[Glibc]
version = 2.24
```
```
(gdb) python print(gdb.selected_frame())
{stack=0xffffd5a8,code=0xf7fd7c80,!special}
```


Don't know what's wrong. Look into this

**Comment by thinkycx** [comment]
I have conflicted with this problem twice.Now I have forked this repo and modify it  for ubuntu16.04.
More info you can find here.
https://github.com/thinkycx/libheap

**Comment by cloudburst** [comment]
Hopefully fixed by fa34afce2f81f10eead671d40485d81e92785aab
